firmware/Core/Src/timekeeping/timekeeping.c

```diff
@@ -0,0 +1,62 @@
+#include "timekeeping/timekeeping.h"
+#include "stm32l4xx_hal.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+
+uint64_t TIM_unix_epoch_time_at_last_time_resync_ms = 0;
+uint32_t TIM_system_uptime_at_last_time_resync_ms = 0;
+TIM_sync_source_t TIM_last_synchroization_source = TIM_SOURCE_NONE;
+
+uint32_t TIM_get_current_system_uptime_ms(void) {
+    return HAL_GetTick();
+}
+
+/// @brief Use this function in a telecommand, or upon receiving a time update from the GPS. 
+void TIM_set_current_unix_epoch_time_ms(uint64_t current_unix_epoch_time_ms, TIM_sync_source_t source) {
+    TIM_unix_epoch_time_at_last_time_resync_ms = current_unix_epoch_time_ms;
+    TIM_system_uptime_at_last_time_resync_ms = HAL_GetTick();
+    TIM_last_synchroization_source = source;
+}
+
+/// @brief Returns the current unix timestamp, in milliseconds
+uint64_t TIM_get_current_unix_epoch_time_ms() {
+    return (HAL_GetTick() - TIM_system_uptime_at_last_time_resync_ms) + TIM_unix_epoch_time_at_last_time_resync_ms;
+}
+
+/// @brief Returns a human-friendly timestamp string for logging. The string
+/// identifies the current Unix time, and number of milliseconds since the
+/// time was last synchronized to a reference source.
+void TIM_get_log_timestamp_string(char *log_str, size_t max_len) {
+    uint64_t delta_uptime = TIM_get_current_system_uptime_ms() - TIM_system_uptime_at_last_time_resync_ms;
+    uint64_t epoch = TIM_get_current_unix_epoch_time_ms();
+
+    time_t seconds = (time_t)(epoch/ 1000U);
+    uint16_t ms = epoch - 1000U * seconds;
+    struct tm *time_info = gmtime(&seconds);
+    char source;
+    switch (TIM_last_synchroization_source) {
+        case TIM_SOURCE_GNSS:
+            source = 'G';
+            break;
+        case TIM_SOURCE_TELECOMMAND:
+            source = 'T';
+            break;
+        default:
```

Comment:
Just noticed it, but each time source enum value should be written out explicitly, and the `default` should ideally throw an error (or in this case, where there is no error return, should do something like `source = 'E';` (for "Error") or `source = '?';`

In a "rare" (but by Murphy's Law, probable) case where we add a new time source enum value, but forget to update this function, we'd get unexpected behaviour where both "no time source" and whatever the new enum value is are represented by 'N'

Reply:
Great catch! The character is now determined in a separate function:

```
char TIM_synchronization_source_letter(TIM_sync_source_t source) {

    char source_letter;
    switch (TIM_last_synchronization_source) {
        case TIM_SOURCE_GNSS:
            source_letter = 'G';
            break;
        case TIM_SOURCE_TELECOMMAND:
            source_letter = 'T';
            break;
        case TIM_SOURCE_NONE:
            source_letter = 'N';
            break;
        default:
            source_letter = 'E';
            break;
    }

    return source_letter;
}
```